### PR TITLE
Implemented keras for tensorkart in python3.5

### PR DESCRIPTION
**Pull request by sirvincent**
First up thank you very much for TensorKart!! I am enjoying playing around with the software.

I expanded your code to work in python 3.5, only slight changes needed to be made :); print statements with () and different module: used the newly released wxPython4.0.01a as wx.
Furthermore I implemented Keras2.0.3 to work with the software, added as two single codes play_keras/train_keras. Compared to tensorflow you can see how much easier it is to make a CNN in Keras. It is important to note that the keras model is in a function in the train_keras.py file not in a seperate file. 

I tried multiple things to make the neural network a better driver but to no avail. Some things I tried: 
1) more training data. The network still barely can drive on a new map. Probably due to the jerkiness of the training data. Trained on mario/luigi raceway and kalimari desert. (each track trained 3 times on and also tested with 3 times training on mario raceway and 2 times the other two.)
2) randomly shuffling the training images. This has as the pro that we can use a validation set during training. From which I found that after training for ~15 epochs the loss on it did not decrease.
3)  only train on 1 road type. Same result as point 1
4) remove top part of the training image because that is something which we also do not use during simple driving (see utils.py commented code). Looks as if it became slightly better but hard to tell 
5) the first three layers of the CNN larger. No noticeable difference
6)  as final layer activation softsign and not tanh. Softsign looks better but again no noticeable difference. 
7) different loss function than the L2 (MSE); euclidean distance between input and output joystick, I think this will punish differences harder.

All in all sometimes a trained network is capable of completely going around 3 laps on mario raceway which is part of the TRAINING tracks but it was never able to finish a lap on royal raceway. In fact it always had problems with the first corners. The part of the track you made a gif of it most of the time did not succeed in staying on track. I think the main problem is that my own training data is more jerky than yours.

Was your network possible of completing a lap on an untrained track?
Can you maybe send me your training data? I think you did a much better job at driving than me.
Do you maybe know how NVIDIA came up with their specific neural network?

Currently I am still playing around with the software and thinking of three things of implementing:
1) acquire more data and test with different image sizes (use e.g. only the track as image)
2) a different neural network e.g.: https://github.com/Sentdex/pygta5 imagenet 
 or maybe a lstm, or even a self-learning network
3) computer vision for recognition of the track. One cool thing I would like to try is to let the network race on toads turnpike and dodge the traffic

I am still learning git, so if there is anything wrong please let me know!

### CODE REVIEW

**Comment by kevinhughes27** [comment]
First thanks for this work :heart: I had listed changing to keras as a project I was hoping someone would take on. I'm glad to hear you are enjoying playing with this project :smile: 

As for your results they sound pretty decent. To be clear my model could not complete laps on the royal raceway either. It also failed on the first corner but through using manual override I found those other sections that it could handle. With a lot of data I am sure it would be able to finish laps but I wasn't particularly interested in taking things that far. My goal for the project was to make the entire process seem approachable. If this interests you I definitely encourage you to take it further! I've discussed crowd sourcing training data and reinforcement learning as interesting avenues to explore. 

I'd really like to get these changes merged in to TensorKart! That being said I have a few requests about how to do so (we'll level up your git flow skills in the process :smile: )

Essentially it comes down to splitting this work up into several smaller pull requests. From a quick look here is what I would suggest for patches:
1. PR to upgrade everything required for python3
2. Upgrade wx version (hopefully this fixes the issues lots of others are having with wx)
3. Change to keras. I don't see any need to keep the old code around after this. I'd prefer we just have keras going forwards
4. Any other updates to the training process or model

Let me know if that makes sense! I am excited to get these changes in :rocket:

**Comment by millerda** [comment]
As for switching over totally to keras, I know that the TensorFlow roadmap is actually to include most of the same functionality pretty soon with implementations of popular layers and out of the box networks using best practices. In fact, it seems a lot of the keras functionality is already included in the [tf.contrib.keras](https://www.tensorflow.org/api_docs/python/tf/contrib/keras) module.

**Comment by kevinhughes27** [comment]
Yes keras has been merged into TensorFlow (or at least announced that they are going to do this). This code is still valid though since we still need to re-write the model to make use of keras regardless of whether we include the library from TensorFlow or from standalone.

**Comment by kevinhughes27** [comment]
Can we close this pull request now? I think all the changes were merged separately right?

**Comment by sirvincent** [comment]
Yes we can close this pull request. All changes were merged separately
